include/nil/blueprint/components/systems/snark/plonk/placeholder/permutation_argument_verifier.hpp

```diff
@@ -0,0 +1,359 @@
+                typename BlueprintFieldType::value_type q_last_y = var_value(assignment, instance_input.q_last);
+                typename BlueprintFieldType::value_type q_pad_y = var_value(assignment, instance_input.q_pad);
+
+                for (std::size_t i = 0; i < m; i++) {
+                    fe = fe * (f[i] + theta_1 * Se[i] + theta_2);
+                    fsigma = fsigma * (f[i] + theta_1 * Ssigma[i] + theta_2);
+                    assignment.witness(component.W(0), row + i) = fe;
+                    assignment.witness(component.W(1), row + i) = f[i];
+                    assignment.witness(component.W(2), row + i) = Se[i];
+                    assignment.witness(component.W(4), row + i) = Ssigma[i];
+                    assignment.witness(component.W(5), row + i) = fsigma;
+
+                    if (i & 1) {
+                        assignment.witness(component.W(3), row + i) = theta_2;
+                    } else {
+                        assignment.witness(component.W(3), row + i) = theta_1;
+                    }
+                }
+                row += component.m;
+
+                assignment.witness(component.W(0), row) = L0_y * (one - Vsigma_y);
+                assignment.witness(component.W(1), row) = q_last_y;
+                assignment.witness(component.W(2), row) = q_pad_y;
+                assignment.witness(component.W(3), row) = L0_y;
+                assignment.witness(component.W(4), row) =
+                    (1 - (q_last_y + q_pad_y)) * (Vsigma_zetay * fsigma - Vsigma_y * fe);
+
+                row++;
+
+                assignment.witness(component.W(0), row) = q_last_y * (Vsigma_y * Vsigma_y - Vsigma_y);
+                assignment.witness(component.W(1), row) = Vsigma_y;
+                assignment.witness(component.W(2), row) = Vsigma_y * Vsigma_y;
+                assignment.witness(component.W(3), row) = Vsigma_zetay;
+
+                return typename plonk_permutation_verifier<BlueprintFieldType, ArithmetizationParams>::result_type(
+                    component, start_row_index);
+            }
+
+            template<typename BlueprintFieldType, typename ArithmetizationParams>
+            std::vector<std::size_t> generate_gates(
+                const plonk_permutation_verifier<BlueprintFieldType, ArithmetizationParams> &component,
+                circuit<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &bp,
+                assignment<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>>
+                    &assignment,
+                const typename plonk_permutation_verifier<BlueprintFieldType, ArithmetizationParams>::input_type
+                    instance_input) {
+
+                using var = typename plonk_permutation_verifier<BlueprintFieldType, ArithmetizationParams>::var;
+
+                auto constraint_1 = var(component.W(0), 0) - var(component.W(1), 0) -
```

Comment:
Please rename the variables to their meanings for each gate.
It is easier to check the mapping once and understand the constraints, than to rebuild the mapping by reading the assignments and check each time.